Fix #4445: Add scaling animation for Continue button

JishnuGoyal opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #4445

This PR is part of the GSoC project: Interactive Onboarding Flow which fixes #4445. It does so by creating a new custom view which connects to the `ContinueInteractionItemViewModel`, which receives a flag to start animating from the explorationProgressController. We wait for 45 seconds when the first card of a lesson is opened by the user, and then start the animation for the continue button. 

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing


https://user-images.githubusercontent.com/64526117/202579860-643e2d20-b969-438d-be86-f690630b8278.mp4

JishnuGoyal committed:
add EnableExtraTopicTabsUi platform parameter

JishnuGoyal committed:
use platform parameter to hide extra topic tabs

JishnuGoyal committed:
fix errors

JishnuGoyal committed:
shorten line length

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt

JishnuGoyal committed:
add force method

JishnuGoyal committed:
make changes to the test. Not completed yet.

JishnuGoyal committed:
reset to develop

JishnuGoyal committed:
add tests

JishnuGoyal committed:
update to correct logic

JishnuGoyal committed:
change parameter name

JishnuGoyal committed:
fix some incorrect code

JishnuGoyal committed:
fix some incorrect code

JishnuGoyal committed:
adjust code changes for tests in this file

JishnuGoyal committed:
remove unused resources

JishnuGoyal committed:
lint

JishnuGoyal committed:
merge with develop

JishnuGoyal committed:
add platform parameter module to alpha kenya app component

JishnuGoyal committed:
remove unused resource - position with three tabs

JishnuGoyal committed:
nit: concise kdoc

JishnuGoyal committed:
add accidentally removed modules

JishnuGoyal committed:
update logic ; review required

JishnuGoyal committed:
lint

JishnuGoyal committed:
reset all changes to AlphaKenyaApplicationComponent.kt

JishnuGoyal committed:
add parameter to alpha kenya module

JishnuGoyal committed:
remove "isSuccessful"

JishnuGoyal committed:
add suggested changes

JishnuGoyal committed:
lint

JishnuGoyal committed:
update logic

JishnuGoyal committed:
update test logic to account for the removed tabs

JishnuGoyal committed:
lint

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/topic/TopicFragmentTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentTest.kt

JishnuGoyal committed:
suggested changes

JishnuGoyal committed:
lint

JishnuGoyal committed:
partial: delete practice tab module

JishnuGoyal committed:
add custom view

JishnuGoyal committed:
use expand animation

JishnuGoyal committed:
complete: remove practice tab module and enablePracticeTab

JishnuGoyal committed:
Merge remote-tracking branch 'origin/1.3_hide_topic_tabs' into 1.3_hide_topic_tabs

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

JishnuGoyal committed:
add flag to ephemeral state

JishnuGoyal committed:
lint

JishnuGoyal committed:
update logic

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

# Conflicts:
#	app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt
#	app/src/sharedTest/java/org/oppia/android/app/topic/revision/TopicRevisionFragmentTest.kt

JishnuGoyal committed:
fix code after a bad merge

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

JishnuGoyal committed:
update some tests according to changes introduced due to hiding tabs

JishnuGoyal committed:
lint

JishnuGoyal committed:
update test due to hiding some tabs

JishnuGoyal committed:
lint

JishnuGoyal committed:
update tests

JishnuGoyal committed:
update rest of the tests to include changes from hiding tabs + modify chapter list PR

JishnuGoyal committed:
lint

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

JishnuGoyal committed:
all functionality added - not working

ben's suggestions

commit before checkout

JishnuGoyal committed:
dont use delegates

JishnuGoyal committed:
lint

JishnuGoyal commented on domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:
```diff
@@ -10,21 +10,25 @@ import org.oppia.android.app.model.PendingState
  * treats the learner's progress like a deck of cards to simplify forward/backward navigation.
  */
-class StateDeck constructor(
+class StateDeck @Inject constructor(
```
should be removed

JishnuGoyal replied:
done

JishnuGoyal commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -117,6 +118,8 @@ class ExplorationProgressController @Inject constructor(
   //  for getCurrentState).
   private lateinit var profileId: ProfileId
+  var timestampToAnimate: Long = -1
+  val isContinueButtonAnimating = MutableLiveData(false)
```
should be removed

JishnuGoyal replied:
done

JishnuGoyal commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -36,7 +40,7 @@
     app:questionViewMarginTop="@{@dimen/general_button_item_margin_top}">
 
-  <ImageButton
+    <ImageButton
```
code formatting is not correct

JishnuGoyal replied:
done

JishnuGoyal commented on app/src/main/res/layout/animating_continue_interaction_item.xml:
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="utf-8"?>
```
this file can be deleted

JishnuGoyal replied:
done

JishnuGoyal commented on app/src/main/res/anim/expand.xml:
```diff
@@ -0,0 +1,9 @@
+    android:fromYScale="1.0"
+    android:toXScale="2.0"
+    android:toYScale="2.0" />
+</set>
```
add eof newline

JishnuGoyal replied:
done

JishnuGoyal commented on app/src/main/java/org/oppia/android/app/player/state/StatePlayerRecyclerViewAssembler.kt:
```diff
@@ -1222,7 +1222,11 @@ class StatePlayerRecyclerViewAssembler private constructor(
         viewType = StateItemViewModel.ViewType.CONTINUE_INTERACTION,
         inflateDataBinding = ContinueInteractionItemBinding::inflate,
-        setViewModel = ContinueInteractionItemBinding::setViewModel,
+        setViewModel = { continueInteractionItemBinding, viewModel ->
```
should be removed as this change is no longer needed

JishnuGoyal replied:
done

JishnuGoyal commented on app/src/main/java/org/oppia/android/app/player/state/StateFragmentPresenter.kt:
```diff
@@ -332,6 +333,34 @@ class StateFragmentPresenter @Inject constructor(
+//    setUpContinueButtonAnimation(ephemeralState)
+  }
+
+  private fun setUpContinueButtonAnimation(ephemeralState: EphemeralState) {
```
these changes are no longer needed

JishnuGoyal replied:
done

JishnuGoyal reviewed (commented):
This is a self review pass before sending in a formal review request.

JishnuGoyal committed:
delete unused files

JishnuGoyal committed:
pass timestamp instead of oppia clock

JishnuGoyal committed:
add isVisibilityAggregated -- solves the issue of animating when view is off screen.

JishnuGoyal committed:
all working.

JishnuGoyal committed:
rebuild app

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove unused variables and code cleanup

JishnuGoyal commented:
PTAL @BenHenning please take an initial pass at this. Tests not completed at the moment.

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,54 @@
+    viewModel.animateContinueButton.observeForever(continueButtonAnimationObserver)
+  }
+
+  private val continueButtonAnimationObserver = Observer<Boolean> {
```
Nit on mixing fields & methods--generally fields should be organized together, and ditto for methods, rather than mixing them.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,54 @@
+  fun setViewModel(viewModel: ContinueInteractionViewModel) {
+    this.viewModel = viewModel
+    viewModel.subscribeToCurrentState()
+    viewModel.animateContinueButton.observeForever(continueButtonAnimationObserver)
```
We ought to never use observeForever--one advantage we have performing observation at this level is having a lifecycle owner (the parent fragment) that can be used.

JishnuGoyal replied:
the parent fragment hasn't been injected into the custom view .... (and probably cannot be? am i missing something?). 
One option here is to get a reference to the parent fragment from the viewmodel, but that seems to  be an anti pattern?

BenHenning replied:
If you set the view up for injection, you should be able to inject its parent fragment.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,54 @@
+    // timer goes off. To make sure that the animation is shown when the user finally scrolls down
+    // enough that the button is visible, animate the button based on whether the timer has finished
+    // or not.
+    if (isVisible && isAnimationTimerFinished) {
```
When does this happen? ``isAnimationTimerFinished`` is set to true when ``startAnimating`` is called, so this seems a bit backwards.

JishnuGoyal replied:
changed the position of this line

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/ContinueInteractionViewModel.kt:
```diff
@@ -45,9 +57,51 @@ class ContinueInteractionViewModel private constructor(
+      }
+//        oppiaLogger.e("StateFragment", "Failed to retrieve ephemeral state", result.error)
+      is AsyncResult.Pending -> {
+      } // Display nothing until a valid result is available.
```
Reminder to finish these bits.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,54 @@
+
+  private fun startAnimating() {
+    val animation = AnimationUtils.loadAnimation(context, R.anim.expand)
+    animation.interpolator = AccelerateInterpolator()
+    animation.repeatCount = Int.MAX_VALUE
```
Can't these be defined as part of the animation itself?

JishnuGoyal replied:
I'm not sure why but for the scaling animation, including this in the animation file is not possible:
![image](https://user-images.githubusercontent.com/64526117/192156822-f624f518-13fe-4a69-bf5c-b1a159352370.png)

JishnuGoyal replied:
DONE.

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  private fun startAnimating() {
+    val animation = AnimationUtils.loadAnimation(context, R.anim.expand)
```
We should name the animation something more contextually specific to what it's being used for (e.g. maybe "pulse_button_size" or something similar?).

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -46,14 +50,19 @@
-    <Button
-      android:id="@+id/continue_button"
+    <org.oppia.android.app.customview.ContinueButtonView
+      android:id="@+id/animation_continue_button"
```
Why change the ID?

JishnuGoyal replied:
appropriate name added

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -6,14 +6,18 @@
   <androidx.constraintlayout.widget.ConstraintLayout
     android:layout_width="match_parent"
-    android:layout_height="wrap_content"
+    android:layout_height="50dp"
```
This seems a bit arbitrary--why 50?

JishnuGoyal replied:
this change was removed

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -46,14 +50,19 @@
+      android:layout_marginEnd="10dp"
+      android:gravity="center"
+      android:minWidth="96dp"
+      android:minHeight="36dp"
```
Button min sizes should always be 48dp.

That being said, why are these new min width/heights needed?

JishnuGoyal replied:
For this and some of the further comments: Changing the view type from Button to CustomButtonView messes with the layout design (adding the same) which is why some of these changes are required. Please let me know if some changes are still needed.
![image](https://user-images.githubusercontent.com/64526117/195989978-c0613113-0bc9-47d6-a4c8-1d013ed069f3.png)

BenHenning replied:
@JishnuGoyal could you perhaps show a before & after to compare? It looks like the text alignment is off.

That being said, do you know what's changing from Button to the custom button view? That sounds like we might be doing something wrong, so I think the goal should be getting the custom view to behave like the original button for consistency, then these extra changes will probably not be needed.

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -46,14 +50,19 @@
+      android:id="@+id/animation_continue_button"
       style="@style/StateButtonActive"
-      android:layout_width="wrap_content"
+      android:layout_marginEnd="10dp"
```
Why is this needed now?

JishnuGoyal replied:
PTAL at the above image

BenHenning replied:
Closing this to consolidate to a single thread that covers "why are there extra changes needed."

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -46,14 +50,19 @@
+      android:minHeight="36dp"
       android:onClick="@{(v) -> viewModel.handleButtonClicked()}"
       android:text="@string/state_continue_button"
+      android:textSize="14sp"
```
Why this change?

BenHenning replied:
Covered in the thread above as part of "why are extra changes needed." Closing this to consolidate the thread.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:
```diff
@@ -518,6 +519,71 @@ class ExplorationActivityTest {
+      )
+
+      onView(withId(R.id.animation_continue_button)).check(matches(not(isAnimating())))
+      testCoroutineDispatchers.advanceTimeBy(45000)
```
This will cause problems for Espresso tests (it's a *really* long wait). We'll probably need to introduce a Dagger configuration value that we can swap out for tests to make sure they don't need to wait the maximum time. See:

https://github.com/oppia/oppia-android/blob/6b58d6a8f677ca67ad13eb2802b864c4a4d0fc02/domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintsAndSolutionConfigModule.kt#L12

and

https://github.com/oppia/oppia-android/blob/6b58d6a8f677ca67ad13eb2802b864c4a4d0fc02/domain/src/main/java/org/oppia/android/domain/hintsandsolution/HintsAndSolutionConfigFastShowTestModule.kt#L8

Alternatively, we could verify the animation only in a local test (i.e. not one in sharedTest). That's probably fine in this case since it isn't *that* important to make sure these run correctly on Espresso (Robolectric should do an okay job here, and we don't yet have a great testing strategy for animations in general).

JishnuGoyal replied:
We aren't using espresso anymore for testing animations. A platform parameter has also been added to disable the animation during espresso tests.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -129,6 +129,8 @@ class ExplorationProgressController @Inject constructor(
   private var mostRecentCommandQueue: SendChannel<ControllerMessage<*>>? = null
 
+  // the amount of time to wait before the continue interaction button is animated in milliseconds
+  private val continueButtonAnimationDelay: Long = 45000L
```
```suggestion
  private val continueButtonAnimationDelayMs: Long = 45000L
```

BenHenning replied:
Also, I suggest using TimeUnit to calculate the value (see my reference module links above for a sample).

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -835,7 +837,10 @@ class ExplorationProgressController @Inject constructor(
-    explorationProgress.stateDeck.getCurrentEphemeralState(retrieveCurrentHelpIndex())
+    explorationProgress.stateDeck.getCurrentEphemeralState(
+      retrieveCurrentHelpIndex(),
+      oppiaClock.getCurrentTimeMs() + continueButtonAnimationDelay
```
This actually doesn't seem correct, and I think will have an issue in the situation where you expand/collapse answers. We need to be setting the initial time (as the session start time) as part of the session initialization. It can be stored as a value within ControllerState like other session-specific data.

JishnuGoyal replied:
done -- PTAL

BenHenning commented on domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt:
```diff
@@ -195,6 +194,9 @@ class StateDeck constructor(
           .addAllWrongAnswer(currentDialogInteractions)
           .setHelpIndex(helpIndex)
       )
+      .setContinueButtonAnimationTimestamp(
+        timestamp
+      )
```
Can fit on one line.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt:
```diff
@@ -615,7 +617,10 @@ class QuestionAssessmentProgressController @Inject constructor(
-    progress.stateDeck.getCurrentEphemeralState(hintHandler.getCurrentHelpIndex().value)
+    progress.stateDeck.getCurrentEphemeralState(
+      hintHandler.getCurrentHelpIndex().value,
+      oppiaClock.getCurrentTimeMs()
```
Should we also verify via a test that questions do NOT animate their continue buttons?

JishnuGoyal replied:
done

BenHenning replied:
This seems unresolved--I don't see any new tests for questions. Did I somehow miss them?

BenHenning replied:
The latest changes have sufficient continue button UI & domain-level tests.

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -270,6 +270,9 @@ message EphemeralState {
   string session_id = 9;
+
+  // The timestamp at which continue button should start animating.
+  int64 continue_button_animation_timestamp = 10;
```
Please include the unit (e.g. ms).

JishnuGoyal replied:
done

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -270,6 +270,9 @@ message EphemeralState {
   // The identifier of the current session that'd be used for analytics logging.
   string session_id = 9;
+
+  // The timestamp at which continue button should start animating.
```
```suggestion
  // The timestamp at which continue buttons (both interaction & navigation) should start animating.
```

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -46,14 +50,19 @@
 
-    <Button
-      android:id="@+id/continue_button"
+    <org.oppia.android.app.customview.ContinueButtonView
```
Shouldn't the navigation button layout & view model also be updated to use this custom view?

JishnuGoyal replied:
done

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal. Took a first pass on the PR & left some various thoughts. I think the approach makes sense. Much of my comments are either pointing out things that may have been missed, or looking for clarification into specifics of the approach (to make sure that they're actually necessary).

JishnuGoyal committed:
add platform parameter to disable animation

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove duplicate entry and revert the "reversion"

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -51,4 +64,16 @@ class ContinueButtonView @JvmOverloads constructor(
+        FragmentManager.findFragment<Fragment>(this) as ViewComponentFactory
+      val viewComponent = viewComponentFactory.createViewComponent(this) as ViewComponentImpl
+      viewComponent.inject(this)
+    } catch (e: IllegalStateException) {
```
we shouldn't need to catch this. Why is it needed?

JishnuGoyal replied:
removing the try catch block

BenHenning commented on app/src/main/java/org/oppia/android/app/view/ViewComponentImpl.kt:
```diff
@@ -31,6 +32,7 @@ interface ViewComponentImpl : ViewComponent {
   fun inject(dragDropSortInteractionView: DragDropSortInteractionView)
   fun inject(imageRegionSelectionInteractionView: ImageRegionSelectionInteractionView)
   fun inject(lessonThumbnailImageView: LessonThumbnailImageView)
+  fun inject(continueButtonView: ContinueButtonView)
```
These injections should be in alphabetical order--please update.

JishnuGoyal replied:
done

BenHenning commented on utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt:
```diff
@@ -110,6 +110,17 @@ const val ENABLE_PERFORMANCE_METRICS_COLLECTION = "enable_performance_metrics_co
+/**
+ * Qualifier for the platform parameter that controls whether to animate the continue button
+ * interaction and navigation items. This is used to disable the animation during testing because
+ * espresso has known problems while testing views that contain animations.
```
```suggestion
 * Espresso has known problems while testing views that contain animations.
```

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -4,16 +4,27 @@ import android.content.Context
   defStyleAttr: Int = R.style.StateButtonActive
 ) : androidx.appcompat.widget.AppCompatButton(context, attrs, defStyleAttr) {
 
+  @Inject
+  @EnableContinueButtonAnimation
+  lateinit var enableContinueButtonAnimation: PlatformParameterValue<Boolean>
```
```suggestion
  @field:[Inject EnableContinueButtonAnimation]
  lateinit var enableContinueButtonAnimation: PlatformParameterValue<Boolean>
```

You probably need to do this since it's field injection.

JishnuGoyal replied:
done

BenHenning reviewed (commented):
Thanks @JishnuGoyal. Took a quick incremental pass & left a few comments. I think I might've discovered the reason for your Gradle build failures.

JishnuGoyal committed:
change position of function call to a more appropriate line

JishnuGoyal committed:
Merge remote-tracking branch 'origin/develop' into develop

JishnuGoyal committed:
Merge branch 'develop' into 1.3_hide_topic_tabs

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove unused resource

JishnuGoyal committed:
mend incorrect legacy logic

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt

JishnuGoyal committed:
lint

JishnuGoyal committed:
partial: review changes

JishnuGoyal committed:
partial: review changes

JishnuGoyal committed:
remove usages of delegates

JishnuGoyal committed:
lint

JishnuGoyal committed:
lint

JishnuGoyal committed:
platform parameter was not properly defined

JishnuGoyal committed:
wrong module was being used

JishnuGoyal committed:
Merge branch '1.3_hide_topic_tabs' into continue_button_animation

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterAlphaKenyaModule.kt
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt
#	testing/src/main/java/org/oppia/android/testing/platformparameter/TestPlatformParameterModule.kt

JishnuGoyal committed:
merge with develop

JishnuGoyal committed:
add bazel dep

JishnuGoyal committed:
add bazel & remove unused resource fron topicListController

JishnuGoyal committed:
commit before checkout

JishnuGoyal committed:
change delay to 45s

JishnuGoyal committed:
shift tests

JishnuGoyal committed:
remove dependency on the view model for continue btn anim

JishnuGoyal committed:
update tests

JishnuGoyal committed:
Merge branch 'remove_dependency_on_view_models_for_anim' into continue_button_animation

JishnuGoyal committed:
partial: support for per profile setting

JishnuGoyal committed:
all functionality working

JishnuGoyal committed:
lint

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

JishnuGoyal committed:
lint

JishnuGoyal commented:
PTAL @BenHenning

BenHenning commented on app/BUILD.bazel:
```diff
@@ -160,6 +160,7 @@ LISTENERS = [
     "src/main/java/org/oppia/android/app/topic/lessons/StorySummarySelector.kt",
     "src/main/java/org/oppia/android/app/topic/revision/RevisionSubtopicSelector.kt",
     "src/main/java/org/oppia/android/app/topic/revisioncard/ReturnToTopicClickListener.kt",
+    "src/main/java/org/oppia/android/app/utility/LifecycleSafeTimerFactory.kt",
```
This isn't a listener--it shouldn't be in this list.

It's also already in the build graph which means this is duplicating its presence in the build. You may need to update deps if you need this for a target that doesn't have access to it.

JishnuGoyal replied:
removed

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -51,7 +52,7 @@
 
     <org.oppia.android.app.customview.ContinueButtonView
-      android:id="@+id/animation_continue_button"
+      android:id="@+id/continue_button"
```
Can this be more specific? View IDs are global in the app.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt:
```diff
@@ -619,7 +619,8 @@ class QuestionAssessmentProgressController @Inject constructor(
       hintHandler.getCurrentHelpIndex().value,
-      oppiaClock.getCurrentTimeMs()
+      oppiaClock.getCurrentTimeMs(),
+      true
```
This needs to include the parameter name to provide context as to what the ``true`` value corresponds to, e.g. ``parameterName = true``

JishnuGoyal replied:
done

BenHenning commented on model/src/main/proto/exploration.proto:
```diff
@@ -273,6 +273,9 @@ message EphemeralState {
   int64 continue_button_animation_timestamp = 10;
+
+  // Determines if the continue button animation should be shown or not.
+  bool show_continue_button_animation = 11;
```
Nit: swap this & the above's places and indexes (it makes more sense, I think, for the bool to go before the long since it "gates" whether to do the action at all).

JishnuGoyal replied:
done

BenHenning commented on model/src/main/proto/profile.proto:
```diff
@@ -73,6 +73,9 @@ message Profile {
   // Represents the number of times a user has successfully logged into their profile.
   int32 number_of_logins = 14;
+
+  // Determines if the continue button animation has been seen by the user or not.
```
This needs a bit more context since it's at the profile level, not at the exploration level, contextually. Suggest:

```suggestion
  // Determines if the in-exploration continue button animation has been seen by the user or not.
```

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,129 @@
+    }
+  }
+
+  private fun processEphemeralStateResult(result: AsyncResult<EphemeralState>) {
```
This needs to be finished.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,129 @@
+
+  private fun processEphemeralState(ephemeralState: EphemeralState) {
+    if (!ephemeralState.showContinueButtonAnimation) return
+    if (!ephemeralState.hasPreviousState) {
```
Why can't this logic be in the ExplorationProgressController? It seems more complicated to have some of the conditions to show the animation there, and some of them here. It should be cleaner just to have the view show the animation when instructed and otherwise require no additional logic.

JishnuGoyal replied:
does exploration progress controller have access to the ephemeral state? I suppose state deck has it, but we arent injecting it into the continue button view

JishnuGoyal replied:
done -- it was added to the function where ephemeralState is being initialized

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,129 @@
+  private lateinit var viewModel: ContinueInteractionViewModel
+  private var isAnimationTimerFinished = false
+
+  val animateContinueButton = MutableLiveData(false)
```
Can this not be an ObservableBoolean, instead?

JishnuGoyal replied:
this needs to be a live data, as its value is observed for triggering an animation

BenHenning commented on app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/ContinueInteractionViewModel.kt:
```diff
@@ -65,7 +63,7 @@ class ContinueInteractionViewModel private constructor(
         fragment as PreviousNavigationButtonListener,
         isSplitView,
-        writtenTranslationContext
+        writtenTranslationContext,
```
```suggestion
        writtenTranslationContext
```

Small nit for consistency.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/anim/expand.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Reminder that this file needs to be named more specifically.

Also, have you adjusted this animation to better match the mocks?

JishnuGoyal replied:
I have verified with @mschanteltc that the current animation looks good.

BenHenning commented on app/src/main/res/values/ids.xml:
```diff
@@ -5,4 +5,5 @@
   <item name="onboarding_dot_1" type="id" />
   <item name="onboarding_dot_2" type="id" />
   <item name="onboarding_dot_3" type="id" />
+  <item name="continue_button" type="id" />
```
Why is this needed?

BenHenning replied:
It seems that this has been removed in the latest changes.

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -267,6 +268,67 @@ class StateFragmentLocalTest {
+      testCoroutineDispatchers.runCurrent()
+
+      onView(withId(R.id.state_recycler_view)).perform(scrollToViewType(CONTINUE_INTERACTION))
+      onView(withId(R.id.continue_button)).check(matches(Matchers.not(isAnimating())))
```
Nit: you should be directly importing ``not``. Ditto elsewhere in the test.

JishnuGoyal replied:
done

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -267,6 +268,67 @@ class StateFragmentLocalTest {
+      onView(isRoot()).perform(orientationLandscape())
+      testCoroutineDispatchers.runCurrent()
+      scrollToViewType(CONTINUE_INTERACTION)
+      onView(withId(R.id.continue_button)).check(matches(Matchers.not(isAnimating())))
+      testCoroutineDispatchers.advanceTimeBy(15000)
+      onView(withId(R.id.continue_button)).check(matches(isAnimating()))
```
This indicates two different tests--I suggest splitting it up, e.g.:

- Doesn't animate after rotating with a 30 second wait
- Does animate if another 15 seconds is waited

You should also remove/split out the check above. Each test should *only* have checks at the end of the test, not at the middle.

JishnuGoyal replied:
done

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -267,6 +268,67 @@ class StateFragmentLocalTest {
+    }
+  }
+
+  private fun isAnimating(): TypeSafeMatcher<View> {
+    return ActiveAnimationMatcher()
+  }
+
+  private class ActiveAnimationMatcher() : TypeSafeMatcher<View>() {
+    override fun describeTo(description: Description) {
+      description.appendText("View is animating")
+    }
+
+    override fun matchesSafely(view: View): Boolean {
+      return view.animation?.hasStarted() ?: false
+    }
+  }
```
Please move these to be below the other tests where other helpers go.

JishnuGoyal replied:
done

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -267,6 +268,67 @@ class StateFragmentLocalTest {
+  }
+
+  @Test
+  fun testConIntAnim_openFractions_expId1_checkButtonDoesNotAnimate() {
```
You also need tests for:
- Verifying that the continue navigation button also animates when expected/doesn't when expected
- Verifying that both the navigation & interaction continue buttons don't animate if they've already been seen in a *new* exploration play session
- Verifying that waiting 30 seconds in one state, then moving to the next and showing its continue button, then waiting another 15 seconds, does NOT show the animation (I think it will today in your current implementation due to you not forcing the animation to not show in the progress controller--I left a relevant comment in that file as well)
- Ideally, verifying that for a separate profile that the animation is seen even if it's been seen by another profile on the device

JishnuGoyal replied:
1) Regarding the continue navigation button animation test (bullet point 1):
There is no way to test this right now, as we dont have any exploration where the "first card" has a navigation item. We do have cards where there is a "submit" button, but pressing submit leads to endState() which causes all animation logic to be suspended for the further cards in the exploration. 

2) Test was added for point 3

BenHenning replied:
Why doesn't (1) also affect non-test cases (i.e. when actually using the app)? Have you tested to make sure that the continue navigation button also animates when expected? If it doesn't, that sounds like something that needs to be fixed (at which point it should become testable if we can manually demonstrate it).

BenHenning replied:
The latest changes have sufficient continue button UI & domain-level tests.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -129,6 +129,8 @@ class ExplorationProgressController @Inject constructor(
 
   private var mostRecentCommandQueue: SendChannel<ControllerMessage<*>>? = null
 
+  // the amount of time to wait before the continue interaction button is animated in milliseconds
```
```suggestion
  // The amount of time to wait before the continue interaction button is animated in milliseconds.
```

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -104,7 +104,7 @@ class ExplorationProgressController @Inject constructor(
   private val dataProviders: DataProviders,
   private val loggingIdentifierController: LoggingIdentifierController,
-  private val profileManagementController: ProfileManagementController,
+  val profileManagementController: ProfileManagementController,
```
Why is this been changed to public? That shouldn't be needed.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -619,6 +627,10 @@ class ExplorationProgressController @Inject constructor(
     }
   }
 
+  fun markContinueButtonAnimationSeen() {
```
This isn't needed since the controller calls into profile management controller itself below.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -408,7 +412,11 @@ class ExplorationProgressController @Inject constructor(
                   learnerId,
-                  learnerAnalyticsLogger
+                  learnerAnalyticsLogger,
+                  oppiaClock.getCurrentTimeMs(),
```
```suggestion
                  startSessionTimeMs = oppiaClock.getCurrentTimeMs(),
```

For clarity.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -408,7 +412,11 @@ class ExplorationProgressController @Inject constructor(
-                  learnerAnalyticsLogger
+                  learnerAnalyticsLogger,
+                  oppiaClock.getCurrentTimeMs(),
+                  isContinueButtonAnimationSeen,
```
It'd probably be cleaner to do the ``?: false`` here and avoid needing to propagate a nullable boolean.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -1106,6 +1126,9 @@ class ExplorationProgressController @Inject constructor(
     fun endState() {
       stateAnalyticsLogger?.logEndCard()
       explorationAnalyticsLogger.endCard()
+      isContinueButtonAnimationSeen?.let {
```
So I'd expect this to be a var, actually, since it needs to be set to false at endState to avoid showing the animation after the first state.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -746,6 +748,39 @@ class ProfileManagementController @Inject constructor(
   }
 
+  /**
+   * Returns true if the continue button animation corresponding to the specified [profileId],
+   * and null if the specified profile doesn't exist.
```
```suggestion
   * Returns whether the exploration continue button animation has shown (or been disabled) for the specified [profileId], or null if the profile doesn't exist.
```

(Needs line wrapping).

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -746,6 +748,39 @@ class ProfileManagementController @Inject constructor(
+
+  /** Marks that the continue button animation has been seen. */
+  fun markContinueButtonAnimationSeen(profileId: ProfileId): DataProvider<Any?> {
+    val deferred = profileDataStore.storeDataWithCustomChannelAsync(
```
Per our discussion, this should not be using the custom channel (it isn't necessary), and we should be calling await() on the deferred rather than converting it to a DataProvider.

JishnuGoyal replied:
done

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal. I took another pass. Please keep in mind you also need tests for ExplorationProgressController & ProfileManagementController.

BenHenning commented:
Also, CI checks should be passing & all comments addressed.

JishnuGoyal committed:
add tests

JishnuGoyal committed:
refactor continue button id

JishnuGoyal committed:
swap places (nit)

JishnuGoyal committed:
small nits

JishnuGoyal committed:
move the logic back to use suspend

JishnuGoyal committed:
logic for finishing the state fragment test activity

JishnuGoyal committed:
commit suggestions. Approach 1

JishnuGoyal committed:
update tests

JishnuGoyal committed:
commit before checkout

JishnuGoyal committed:
Revert "commit before checkout"

This reverts commit b5b46f86

JishnuGoyal committed:
commit suggestions

JishnuGoyal committed:
lint

JishnuGoyal committed:
add another test

JishnuGoyal committed:
remove extra resource

JishnuGoyal committed:
build.bazel

JishnuGoyal committed:
refactor variable name

JishnuGoyal committed:
make variable private

JishnuGoyal committed:
add tests, build.bazel error correction

JishnuGoyal committed:
lint

JishnuGoyal committed:
address review comments

JishnuGoyal committed:
lint

JishnuGoyal committed:
remove unnecessary changes

JishnuGoyal committed:
remove unnecessary changes

JishnuGoyal commented:
PTAL @BenHenning A lot has been changed in this PR since the last review, and I feel it would be worth if this underwent another pass, thanks!

Whats left:
1) Add a test to check the buttons do not animate in a _new exploration_
2) Address the CI checks

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,125 @@
+        this.clearAnimation()
+        isAnimationTimerFinished = false
+      }
+    }
```
Given that the live data is completely contained within this file, could we just call start/clear animation when necessary instead of indirecting the logic through a live data? It seems like calling helpers would be simpler.

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -0,0 +1,125 @@
+        isAnimationTimerFinished = true
+      } else {
+        this.clearAnimation()
+        isAnimationTimerFinished = false
```
Perhaps isAtAnimationEnd would be clearer--logically it could be in a finished state and cleared (i.e. it's been finished recently and is no longer running).

JishnuGoyal replied:
done

BenHenning commented on app/src/main/res/layout/continue_interaction_item.xml:
```diff
@@ -6,12 +6,17 @@
     <import type="android.view.View" />
 
+    <variable
+      name="animateContinueButtonLiveData"
```
This seems unused--can it be removed?

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -1124,9 +1150,11 @@ class ExplorationProgressController @Inject constructor(
+            }
           }
-          INDEXTYPE_NOT_SET, null -> {} // Nothing to do here.
+          INDEXTYPE_NOT_SET, null -> {
+          } // Nothing to do here.
```
Here & elsewhere in this file: please revert the format-only changes.

BenHenning replied:
Still relevant--please resolve this. There are multiple cases in this file, so it's probably worth checking the whole PR just to make sure no style-only changes are added.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -1035,7 +1053,11 @@ class ExplorationProgressController @Inject constructor(
-    private val learnerAnalyticsLogger: LearnerAnalyticsLogger
+    private val learnerAnalyticsLogger: LearnerAnalyticsLogger,
+    val startSessionTimeMs: Long,
+    var isContinueButtonAnimationSeen: Boolean?,
```
It seems like this can no longer be null with the latest code--suggest switching to ``Boolean`` and removing the extra ``let`` below.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -129,6 +130,9 @@ class ExplorationProgressController @Inject constructor(
 
   private var mostRecentCommandQueue: SendChannel<ControllerMessage<*>>? = null
 
+  // The amount of time to wait before the continue interaction button is animated in milliseconds.
+  private val continueButtonAnimationDelay: Long = TimeUnit.SECONDS.toMillis(45)
```
Reminder that the changes in this file also need to be tested (in ExplorationProgressControllerTest).

BenHenning replied:
Tests have been added in the latest changes.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/profile/ProfileTestHelper.kt:
```diff
@@ -120,6 +121,14 @@ class ProfileTestHelper @Inject constructor(
+  }
+
+  suspend fun markContinueButtonAnimationSeen(profileId: ProfileId) {
+    return profileManagementController.markContinueButtonAnimationSeen(profileId)
```
Nit: nothing to ``return`` here.

BenHenning replied:
I believe this is still unresolved.

BenHenning replied:
Ditto @JishnuGoyal.

BenHenning replied:
Code seems to be removed in latest changes.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/profile/ProfileTestHelper.kt:
```diff
@@ -120,6 +121,14 @@ class ProfileTestHelper @Inject constructor(
     )
   }
 
+  fun getProfile(profileId: ProfileId): DataProvider<Profile> {
```
Is this helper actually useful? It's just calling forward to the controller (which a test could do). Perhaps it'd be more useful to return the specific property that you're interested in?

BenHenning replied:
I believe this is still unresolved.

BenHenning replied:
Still unresolved @JishnuGoyal?

BenHenning replied:
Code seems to be removed in latest changes.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/profile/ProfileTestHelper.kt:
```diff
@@ -120,6 +121,14 @@ class ProfileTestHelper @Inject constructor(
+    return profileManagementController.getProfile(profileId)
+  }
+
+  suspend fun markContinueButtonAnimationSeen(profileId: ProfileId) {
```
Ditto here--maybe remove the helper since it's not really adding much.

BenHenning replied:
I believe this is still unresolved.

BenHenning replied:
Ditto @JishnuGoyal.

BenHenning replied:
Code seems to be removed in latest changes.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentProgressController.kt:
```diff
@@ -615,7 +617,11 @@ class QuestionAssessmentProgressController @Inject constructor(
   }
 
   private fun ControllerState.computeBaseCurrentEphemeralState(): EphemeralState =
```
Nit: please switch to block notation (we only use expression syntax for one and two liner methods).

BenHenning replied:
Still unresolved.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -746,6 +748,31 @@ class ProfileManagementController @Inject constructor(
+   * Returns whether the exploration continue button animation has shown (or been disabled) for the
+   * specified [profileId], or null if the profile doesn't exist.
+   */
+  suspend fun fetchContinueAnimationSeenStatus(profileId: ProfileId): Boolean? {
```
This & the below methods should have tests in profileManagementController (since they're being exposed as part of the controller's public API).

BenHenning replied:
Ah, is this still incomplete?

Suggest adding:

- testFetchContinueButtonAnimationStatus_invalidProfile_returnsNull
- testFetchContinueButtonAnimationStatus_realProfile_notSeen_returnsFalse
- testFetchContinueButtonAnimationStatus_realProfile_markedAsSeen_returnsTrue
- testFetchContinueButtonAnimationStatus_realProfile_markedAsSeenTwice_returnsTrue
- testFetchContinueButtonAnimationStatus_realProfile_markedAsSeen_inDiffProfile_returnsFalse

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -47,6 +47,8 @@ private const val GET_DEVICE_SETTINGS_PROVIDER_ID =
   "get_device_settings_provider_id"
 private const val ADD_PROFILE_PROVIDER_ID = "add_profile_provided_id"
 private const val UPDATE_NAME_PROVIDER_ID = "update_name_provider_id"
+private const val MARK_CONTINUE_BUTTON_ANIMATION_SEEN_PROVIDER_ID =
```
Not being used--can be removed.

JishnuGoyal replied:
done

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal! Took a full review pass, and checked old comments. Other than the gaps you mentioned in your last comment, and the ones pointed out in comments, everything else is looking quite good. I think after these points are addressed the PR will likely be quite close to being finished.

JishnuGoyal committed:
remove unused resources + refactor variable name

JishnuGoyal committed:
make isContinueButtonAnimationSeen variable non nullable

JishnuGoyal committed:
remove unused resource

JishnuGoyal committed:
remove dependency on animateContinueButton LiveData

JishnuGoyal committed:
move continue button animation seen logic to other function

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

# Conflicts:
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterAlphaKenyaModule.kt
#	domain/src/main/java/org/oppia/android/domain/platformparameter/PlatformParameterModule.kt
#	testing/src/main/java/org/oppia/android/testing/platformparameter/TestPlatformParameterModule.kt

JishnuGoyal committed:
lint

JishnuGoyal committed:
refactor lifecyclesafetimer factory folder to another location: partial

JishnuGoyal committed:
add fields to curretState to make continue nav button animate

JishnuGoyal committed:
lint

JishnuGoyal committed:
lint

JishnuGoyal committed:
bazel

JishnuGoyal committed:
add tests to for continue nav item; failing

JishnuGoyal committed:
bazel config

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

# Conflicts:
#	testing/src/main/java/org/oppia/android/testing/platformparameter/TestPlatformParameterModule.kt

JishnuGoyal committed:
Merge remote-tracking branch 'origin/continue_button_animation' into continue_button_animation

BenHenning committed:
Improve ContinueButtonView's robustness.

In particular, this change:
- Updates ContinueButtonView to take its animation properties from
higher up to avoid potential inconsistencies between its own syncing
with ExplorationProgressController and the one in StateFragmentPresenter
(which could potentially be inconsistent). It also ensures the view
works correctly when used in question practice sessions.
- Ensures ContinueButtonView can be reused across recycles
(specifically, that its animation timer doesn't affect a later reuse of
the view).

Some Bazel build issues were also fixed.

BenHenning commented:
@JishnuGoyal PTAL at https://github.com/JishnuGoyal/oppia-android/pull/3 which I think will address the issues you were running into with getting StateFragmentLocalTest passing. I think it'll also fix a number of the other failures.

You should also sync with the latest develop to make sure your CI results are recent.

JishnuGoyal committed:
Merge pull request #3 from BenHenning/make-continue-button-view-more-robust-against-potential-failures

Improve ContinueButtonView's robustness

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

JishnuGoyal committed:
Merge remote-tracking branch 'origin/continue_button_animation' into continue_button_animation

JishnuGoyal committed:
fix bug

JishnuGoyal committed:
merge with develop

JishnuGoyal committed:
commit before checkout

JishnuGoyal committed:
add tests for exploration progress controller and QuestionAssessmentProgressControllerTest.kt

JishnuGoyal committed:
disable tests for continue navigation item animation

JishnuGoyal committed:
lint

JishnuGoyal committed:
merge with develop

JishnuGoyal commented:
Thanks @BenHenning for the suggestive PR! This helped a lot. Unfortunately, there still are some problems while testing the continue navigation item. Although the app seems to work normally manually, the test cases still are able to hit a certain case where a deviation from the expected behavior is caught. 
As discussed, the continue navigation animation is a "should have" and not a "must have", so given the time constraints and the scope of this PR, I'm creating an issue for this that can be picked up and worked upon. This PR has work which brings us very close to finding a solution, and appropriate corresponding comments/todos/issues will be created.

Thanks!

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

JishnuGoyal committed:
format

JishnuGoyal commented:
PTAL @BenHenning thanks!

JishnuGoyal committed:
Merge remote-tracking branch 'origin/continue_button_animation' into continue_button_animation

BenHenning commented on app/src/test/java/org/oppia/android/app/player/state/StateFragmentLocalTest.kt:
```diff
@@ -370,6 +373,62 @@ class StateFragmentLocalTest {
+        interactionViewId = R.id.math_expression_input_interaction_view
+      )
+      clickSubmitAnswerButton()
+      testCoroutineDispatchers.advanceTimeBy((TimeUnit.SECONDS.toMillis(45)))
```
Nit: please remove extra parentheses here as they're redundant.

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -139,18 +139,30 @@ class ExplorationProgressControllerTest {
-  @Inject lateinit var translationController: TranslationController
-  @Inject lateinit var fakeAnalyticsEventLogger: FakeAnalyticsEventLogger
-  @Inject lateinit var profileManagementController: ProfileManagementController
+  @get:Rule
+  val oppiaTestRule = OppiaTestRule()
+  @Inject
+  lateinit var context: Context
+  @Inject
+  lateinit var explorationDataController: ExplorationDataController
+  @Inject
+  lateinit var explorationProgressController: ExplorationProgressController
+  @Inject
+  lateinit var fakeExceptionLogger: FakeExceptionLogger
+  @Inject
+  lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+  @Inject
+  lateinit var oppiaClock: FakeOppiaClock
+  @Inject
+  lateinit var explorationCheckpointController: ExplorationCheckpointController
+  @Inject
+  lateinit var monitorFactory: DataProviderTestMonitor.Factory
+  @Inject
+  lateinit var translationController: TranslationController
+  @Inject
+  lateinit var fakeAnalyticsEventLogger: FakeAnalyticsEventLogger
+  @Inject
+  lateinit var profileManagementController: ProfileManagementController
```
Nit: please revert this style-only change.

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:
```diff
@@ -111,6 +111,25 @@ class QuestionAssessmentProgressControllerTest {
+    submitCorrectAnswerForQuestion0()
+    moveToNextQuestion()
+    val ephemeralQuestion = waitForGetCurrentQuestionSuccessfulLoad()
+    assertThat(ephemeralQuestion.ephemeralState.showContinueButtonAnimation).isEqualTo(false)
```
```suggestion
    assertThat(ephemeralQuestion.ephemeralState.showContinueButtonAnimation).isFalse()
```

Ditto elsewhere in this test suite & in ExplorationProgressControllerTest (including for true cases).

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt:
```diff
@@ -111,6 +111,25 @@ class QuestionAssessmentProgressControllerTest {
   }
 
+  @Test
+  fun testEphemeralState_startExploration_shouldIndicateNoButtonAnimation() {
```
Shouldn't be 'exploration'. Please update to match other test names in this test suite.

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -209,6 +221,77 @@ class ExplorationProgressControllerTest {
     assertThat(ephemeralState.state.name).isEqualTo("Continue")
   }
 
+  @Test
+  fun testEphemeralState_startExploration_shouldIndicateButtonAnimation() {
+    oppiaClock.setFakeTimeMode(FakeOppiaClock.FakeTimeMode.MODE_FIXED_FAKE_TIME)
+    val currentTime = oppiaClock.getCurrentTimeMs()
+    oppiaClock.setCurrentTimeMs(currentTime)
+    startPlayingNewExploration(TEST_TOPIC_ID_0, TEST_STORY_ID_0, TEST_EXPLORATION_ID_2)
+    val ephemeralState = waitForGetCurrentStateSuccessfulLoad()
+    assertThat(ephemeralState.showContinueButtonAnimation).isEqualTo(true)
+    assertThat(ephemeralState.continueButtonAnimationTimestampMs).isEqualTo(
+      currentTime + TimeUnit.SECONDS.toMillis(
+        45
+      )
+    )
```
```suggestion
    assertThat(ephemeralState.continueButtonAnimationTimestampMs)
      .isEqualTo(currentTime + TimeUnit.SECONDS.toMillis(45))
```

Not sure what happened here, but the formatting got a bit wonky (plus the equals check can be on its own line).

Ditto elsewhere below.

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt:
```diff
@@ -209,6 +221,77 @@ class ExplorationProgressControllerTest {
+      )
+    )
+  }
+
+  @Test
+  fun testEphemeralState_startExp_clicksContinue_reEnterExp_shouldIndicateNoButtonAnimation() {
+    startPlayingNewExploration(TEST_TOPIC_ID_0, TEST_STORY_ID_0, TEST_EXPLORATION_ID_2)
+    playThroughPrototypeState1AndMoveToNextState()
+    val checkPoint = retrieveExplorationCheckpoint(TEST_EXPLORATION_ID_2)
+    endExploration()
+    resumeExploration(TEST_TOPIC_ID_0, TEST_STORY_ID_0, TEST_EXPLORATION_ID_2, checkPoint)
+    val ephemeralState = waitForGetCurrentStateSuccessfulLoad()
+    assertThat(ephemeralState.showContinueButtonAnimation).isEqualTo(false)
+  }
+
+  @Test
+  fun testEphemeralState_startExp_seesAnimation_reEnterExploration_shouldIndicateButtonAnimation() {
+    oppiaClock.setFakeTimeMode(FakeOppiaClock.FakeTimeMode.MODE_FIXED_FAKE_TIME)
+    startPlayingNewExploration(TEST_TOPIC_ID_0, TEST_STORY_ID_0, TEST_EXPLORATION_ID_2)
+    testCoroutineDispatchers.advanceTimeBy(TimeUnit.SECONDS.toMillis(45))
+    endExploration()
+    val currentTime = oppiaClock.getCurrentTimeMs()
+    oppiaClock.setCurrentTimeMs(currentTime)
```
Here & elsewhere in the code above: why is this needed? Can't we just fetch the time? Not sure I understand why we need to set it back to the same time it started at.

JishnuGoyal replied:
done -- removed

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -488,7 +500,9 @@ class ExplorationProgressController @Inject constructor(
         else -> AsyncResult.Pending()
       }
-    } catch (e: Exception) { AsyncResult.Failure(e) }
+    } catch (e: Exception) {
+      AsyncResult.Failure(e)
+    }
```
Nit: please revert this style-only change.

JishnuGoyal replied:
done

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal. Took another pass and verified old comments. A few old comments haven't yet been resolved, and I noticed some new things to address. It's mostly nits, but a few tests also need to be added I think (I added suggestions for some possible ones to add).

Also, please address the failing CI checks along with these comments before sending the PR back for another review round.

BenHenning commented:
@rt4914 also PTAL for codeowners.

@JishnuGoyal please also update to the latest develop before sending this back for review, as well.

BenHenning commented:
@rt4914 also PTAL for codeowners.

@JishnuGoyal please also update to the latest develop before sending this back for review, as well.

JishnuGoyal committed:
add tests

JishnuGoyal committed:
adjust tests

JishnuGoyal committed:
nits

JishnuGoyal committed:
finishing up

JishnuGoyal committed:
lint

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -63,11 +64,13 @@ class ContinueButtonView @JvmOverloads constructor(
     cancelOngoingTimer()
   }
 
+  /** Sets shouldAnimateContinueButton value. */
```
```suggestion
  /** Sets whether the view should animate to catch a user's attention. */
```

JishnuGoyal replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -63,11 +64,13 @@ class ContinueButtonView @JvmOverloads constructor(
     maybeInitializeAnimation()
   }
 
+  /** Sets continueButtonAnimationTimestampMs value. */
```
```suggestion
  /** Sets when, in clock time, the animation controlled by [setShouldAnimateContinueButton] should play. */
```

(May need wrapping).

JishnuGoyal replied:
done

BenHenning commented on domain/src/test/java/org/oppia/android/domain/profile/ProfileManagementControllerTest.kt:
```diff
@@ -260,6 +260,84 @@ class ProfileManagementControllerTest {
   }
 
+  @Test
+  fun testFetchContinueButtonAnimationStatus_noLoggedInProfile_returnsNull() {
```
Perhaps this should instead test for a profile ID that's not currently logged in? I'm not sure it makes sense to expose separate APIs for the "not logged in" case (see my comment in ProfileManagementController).

JishnuGoyal replied:
done -- test updated for not logged in profile

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -746,6 +746,50 @@ class ProfileManagementController @Inject constructor(
+   *
+   * See [fetchContinueAnimationSeenStatus] for specifics.
+   */
+  suspend fun fetchCurrentContinueAnimationSeenStatus(): Boolean? =
```
Why do we need this? It seems to only be used in testing, so prefer removing it & testing the one that will actually be used (i.e. the one with profile ID being passed in). Ditto for markCurrentContinueButtonAnimationSeen.

JishnuGoyal replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/profile/ProfileManagementController.kt:
```diff
@@ -746,6 +746,50 @@ class ProfileManagementController @Inject constructor(
+    getCurrentProfileId()
+  )
+
+  /** Marks that the continue button animation has been seen. */
```
```suggestion
  /** Marks that the continue button animation has been seen for the specified profile. */
```

JishnuGoyal replied:
done

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -6,6 +6,7 @@ exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/ActivityMo
 exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/ActivityScope.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/InjectableAppCompatActivity.kt"
 exempted_file_path: "app/src/main/java/org/oppia/android/app/activity/route/Route.kt"
+exempted_file_path: "app/src/main/java/org/oppia/android/app/utility/lifecycle/LifecycleSafeTimerFactory.kt"
```
Two things:
1. Please remove the old one (since the file was moved).
2. Please ensure this is in the correct place (per lexicographic ordering).

JishnuGoyal replied:
done

BenHenning reviewed (changes requested):
Thanks @JishnuGoyal. Just had a few follow-up comments, but I think there are still 3 old conversations not yet resolved. Please make sure to resolve these.

Also, CI checks look like they're failing. PTAL.

rt4914 reviewed (changes requested):
@JishnuGoyal I have checked the implementation and also the code and it does look good for code-owner files.

I have one major suggestion:
The animation for continue button is needed so as to bring users attention to it. So I think some similar thing should be done for non-sighted users too. I was thinking when you start the animation for continue button, at that time the screen-reader can speak some thing which brings users attention like "Navigate and click continue button to visit next screen".

@BenHenning Any thoughts?

JishnuGoyal committed:
small corrections

JishnuGoyal committed:
lint

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation

JishnuGoyal committed:
bazel config for profile test helper deps

JishnuGoyal committed:
Merge remote-tracking branch 'origin/continue_button_animation' into continue_button_animation

BenHenning commented on app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt:
```diff
@@ -64,13 +64,14 @@ class ContinueButtonView @JvmOverloads constructor(
   }
 
-  /** Sets continueButtonAnimationTimestampMs value. */
+  /** Sets when, in clock time, the animation controlled by [setShouldAnimateContinueButton] should
+   *  play. */
```
```suggestion
  /**
   * Sets when, in clock time, the animation controlled by [setShouldAnimateContinueButton] should
   * play.
   */
```

Formatting should be consistent & per style guide.

BenHenning reviewed (commented):
Thanks @JishnuGoyal! Just one small nit left--PTAL.

> @JishnuGoyal I have checked the implementation and also the code and it does look good for code-owner files.
> 
> I have one major suggestion: The animation for continue button is needed so as to bring users attention to it. So I think some similar thing should be done for non-sighted users too. I was thinking when you start the animation for continue button, at that time the screen-reader can speak some thing which brings users attention like "Navigate and click continue button to visit next screen".
> 
> @BenHenning Any thoughts?

@rt4914 my only concern with this is that it may change the "default ordering" managed by a screenreader. I think your point is definitely valid if we consider this animation essential to progress (since relying on only visual cues isn't a broadly accessible way of doing things). However, perhaps a cleaner way might be to have a better content description for the continue button since it's more likely that a screenreader user will move through the entire view (and thus not miss something). Given this difference in usage, this may be worth exploring in the future if we get feedback that screenreader users are also getting confused sometimes on their first time through the lesson.

JishnuGoyal committed:
Update app/src/main/java/org/oppia/android/app/customview/ContinueButtonView.kt

nit

Co-authored-by: Ben Henning <henning.benmax@gmail.com>

BenHenning reviewed (approved):
Thanks @JishnuGoyal. LGTM.

rt4914 reviewed (approved):
@JishnuGoyal Approving as @BenHenning suggestion does make sense for A11Y. Thanks.

JishnuGoyal committed:
Merge branch 'develop' into continue_button_animation
